### PR TITLE
Miscellaneous fixes to CPSProjectFactory/CPSProject

### LINKED ISSUE #14306: AbstractProject.LastDesignTimeBuildSucceeded is not always respected
This flags is used to populate ProjectInfo.HasAllInformation, which is used to suppress all errors in the project when design time build is broken for the projects. However, this flag is only read in [CreateProjectInfoForCurrentState](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/AbstractProject.cs,297), which is only invoked when we first start pushing project changes to workspace hosts. Any subsequent changes to this flag have no effect (we are seeing this behavior in CPS).

We should make sure that changing this property after we have started pushing to workspace hosts correctly updates HasAllInformation on the project state.

### LINKED ISSUE #14329: One ProjectExternalErrorReporter is created from 1st project and shared by all projects
**Version Used**: latest preview5

**Steps to Reproduce**:
1. We need to hook up with ProjectExternalErrorReporter to report build errors from DPL.

**Expected Behavior**:
ProjectExternalErrorReporter for each deferred project.

**Actual Behavior**:
One ProjectExternalErrorReporter for all deferred projects:

https://github.com/dotnet/roslyn/blob/49be37c88d52624dd7ec5b4791be0a715426ff2f/src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProjectFactory.cs

       private IVsReportExternalErrors GetExternalErrorReporter(ProjectId projectId, string languageName)
        {
            lock (_externalErrorReporterMap)
            {
                IVsReportExternalErrors errorReporter;
                if **(!_externalErrorReporterMap.TryGetValue(languageName, out errorReporter**))
                {
                    string errorCodePrefix;
                    if (!_projectLangaugeToErrorCodePrefixMap.TryGetValue(languageName, out errorCodePrefix))
                    {
                        throw new NotSupportedException(nameof(languageName));
                    }

                    errorReporter = new ProjectExternalErrorReporter(**projectId**, errorCodePrefix, _serviceProvider);
                    _externalErrorReporterMap.Add(languageName, errorReporter);
                }

                return errorReporter;
            }
        }
@mavasani

### PR DESCRIPTION
**Pull request by mavasani**
1. Make sure that we notify the workspace hosts when IWorkspaceProjectContext.LastDesignTimeBuildSucceeded is updated.
Fixes #14306

2. Ensure that we create an external error reporter per-project ID in the CPSProjectFactory.
Fixes #14329

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @jasonmalinowski @Pilchie @davkean @dotnet/roslyn-ide @dotnet/project-system for review.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject_IIntellisenseBuildTarget.cs [hunk 1]
```diff
@@ -14,29 +14,14 @@ internal partial class AbstractLegacyProject : IIntellisenseBuildTarget
...
-            if (PushingChangesToWorkspaceHosts)
-            {
-                // set workspace reference info
-                ProjectTracker.NotifyWorkspaceHosts(host => (host as IVisualStudioWorkspaceHost2)?.OnHasAllInformation(Id, succeeded));
-            }
+            UpdateIntellisenseBuildFailureDiagnostic(succeeded, reason);
         }
 
-        private void UpdateHostDiagnostics(bool succeeded, string reason)
+        private void UpdateIntellisenseBuildFailureDiagnostic(bool succeeded, string reason)
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject_IIntellisenseBuildTarget.cs]: @davkean - as you had requested, we won't report the intellisense build failure diagnostic for CPS projects as this seemed specific to the native project system. The project system will have to explicitly report their own version of "design time build failed" diagnostic.

**Comment by mavasani** [comment]
test vsi please

**Comment by mavasani** [comment]
Ping.. these are pretty trivial changes and needed on CPS side asap.

**Comment by mavasani** [comment]
test vsi please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_vsi_p0_open_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_vsi_p1_open_prtest please

**Comment by Pilchie** [comment]
:+1:

**Comment by mavasani** [comment]
@dotnet-bot retest windows_vsi_p1_open_prtest please

**Comment by mavasani** [comment]
@dotnet-bot retest windows_vsi_p1_open_prtest please

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject_IIntellisenseBuildTarget.cs (hunk 1 above)
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject_IIntellisenseBuildTarget.cs]: @mavasani: we're still telling CPS projects the yes/no of whether it's overall good, right?

Comment (mavasani) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject_IIntellisenseBuildTarget.cs]: Yes, the flag is still there and will suppress all diagnostics. Design time failure diagnostic is only reported for legacy project system, CPS project system should report its own such diagnostic with appropriate resolution help.

**Pull request merged**